### DISCUSSION TITLE
Relationship Query To Discuss

### DISCUSSION THREAD

**Discussion by rajesh40** [prisma/prisma#12330] [Q&A] [answered]
I have multiple entities like user, customer, company, branch and these all entity have address that is stored in a tables like below
1> addresses -> it conatinse only address having unique id
2> addressbooks  -> this holds relation ship with address and other entities kept here.
addressableType:- if address belongs to any of these like user, customer, branch and company then 
addressableId :-  belongs to id of corresponding user, customer, branch or company

how can we make relation ship in prisma:- 

for more details:- 
```prisma
model Address {
  id         Int      @id @default(autoincrement())
  fname      String?
  lname      String?
  company    String?
  address1   String?
  address2   String?
  address3   String?
  city       String?
  state      String?
  postalCode String?
  country    String?
  phone      String?
  latitude   String?
  longitude  String?
  email      String?
  mobile     String?
  status     Int      @default(1)
  createdAt  DateTime @default(now())
  updatedAt  DateTime @default(now())


  AddressBook AddressBook[]
  @@map("addresses")
}

model AddressBook {
  id              Int             @id @default(autoincrement())
  addressId       Int
  addressableId   Int
  addressableType AddressableType
  type            AddressType
  isDefault       YesNo           @default(no)
  status          Int             @default(1)
  createdAt       DateTime        @default(now())
  updatedAt       DateTime        @default(now())
  address         Address         @relation(fields: [addressId], references: [id])
  User            User?           @relation(fields: [addressableId], references: [id])

  @@map("address_books")
}
```

**Comment by austincrim** [comment, marked as answer]
Hey @rajesh40 👋 ,

The schema you provided looks good so far to me. Could you provide more specifics about which models you have questions about? Could you post your full `schema.prisma` as well?

**Reply by rajesh40** [reply]
users
=========
id 	name
1	Rajesh
2   Ramesh
3 	Jaya


branches
=========
id  name
1	branch1
2   branch2

addresses
=========
id  address1	address2   city
1   my1addd1	my1add2	   my1city
2   my2Add1		my2add2    my2city
3   my3Add1     my3Add2    my3city


addressbooks
==========
id 	addressId	  addressableId		addressableType   
1	1              1                user
2   2              2                user 
3   1              1                branch     
4   2              2                branch  


Now our purpose is,
-- when i will fetch list of users then that data will also contain user address
-- and when i will fetch lis of branches then data will also contain branch address

how can we make relationship in prisma??



```prisma
model Branch {
  id        Int      @id @default(autoincrement())
  name      String
  AddressBook AddressBook[]
  @@map("branches")
}

model User {
  id            Int       @id @default(autoincrement())
  masterId      Int?
  AddressBook       AddressBook[]
  @@unique([userType, email])
  @@map("users")
}

model AddressBook {
  id              Int             @id @default(autoincrement())
  addressId       Int
  addressableId   Int
  addressableType AddressableType
  
  Branch Branch? @relation(fields: [addressableId,addressableType], references: [id])   // line having error due to mismatch parameters
  User   User?   @relation(fields: [addressableId, addressableType], references: [id])  // line having error due to mismatch parameters
  
  @@map("address_books")
}
```
Notice:- 
-- no of columns in [fields] and no of columns in [references] should be same but in our case, it is not.
means in user or branch table there is only id that can belongs here but not

**Reply by rajesh40** [reply]
@austincrim 

main point is .. no of columns in [fields] and no of columns in [references] should be same but in our case, it is not.

if think out of prisma, to get appropriate data we need to make check on address book on two column addressableType and addressableId but in related table like branch, users there is only id column who connects it.

**Reply by austincrim** [reply]
`addressableType` should not be included in the `fields` argument of the `@relation` attribute, I would remove it. You can read more about its purpose [in our docs](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#arguments-6).

You will need to disambiguate your two relations using the `map` arguement since they use the same `addressableId` field.

```prisma
model AddressBook {
  id              Int             @id @default(autoincrement())
  addressId       Int
  addressableId   Int
  addressableType AddressableType

  Branch Branch? @relation(map: "branch_fk", fields: [addressableId], references: [id]) // line having error due to mismatch parameters
  User   User?   @relation(map: "user_fk", fields: [addressableId], references: [id]) // line having error due to mismatch parameters

  @@map("address_books")
}
```

**Reply by rajesh40** [reply]
what about when... branch table and user table both have same id.... ????

means I know, as doc says, my can make relation on multiple field but ..... it's really difficult when addressableId belongs to multiple table and it's id may be same so to manage we have added a  column of addressableType....???

**Reply by rajesh40** [reply]
I will try your suggestion.... then  will let you know??? Thanks man!!!

**Reply by rajesh40** [reply]
```prisma				
model Addressook {
  id              Int             @id @default(autoincrement())
  addressId       Int
  addressableId   Int
  addressableType AddressableType
  type            AddressType
  isDefault       YesNo           @default(no)
  status          Int             @default(1)
  createdAt       DateTime        @default(now())
  updatedAt       DateTime        @default(now())
  address         Address         @relation(fields: [addressId], references: [id])  

  User     User?     @relation(map: "user_fk", fields: [addressableId], references: [id])
  Branch   Branch?   @relation(map: "branch_fk", fields: [addressableId], references: [id])
  Company  Company?  @relation(map: "company_fk", fields: [addressableId], references: [id])
  Location Location? @relation(map: "location_fk", fields: [addressableId], references: [id])
  @@map("address_books")
}
```
```
{"status":false,"data":{},"message":"\nInvalid `prisma.addressBook.create()` invocation:\n\n\n  Foreign key constraint failed on the field: `branch_fk (index)`","payloadError":false}
```

As we know addressbook table is being used for multiple entity table like... company, branch, users etc.
When addressbook is being created for company then for  branch and location  above error is being throws.
while we have marked these relation as optonal , i think optional is not working well.....

**Reply by rajesh40** [reply]
means when in address_book table we are doing entry for company then why it's searching record for other entity while it's optional...

**Reply by austincrim** [reply]
Thanks for the extra detail, @rajesh40, that helps. I think what you are trying to accomplish isn't super well solved by Prisma yet.

I think you should take a look into [this issue](https://github.com/prisma/prisma/issues/2505) for potential solutions/workarounds. 

We also put together [this example repo](https://github.com/prisma/prisma-examples/tree/latest/data-modeling/delegated-types) for doing this type of "polymorphic" relationship.

Let me know if that helps 👍

**Reply by rajesh40** [reply]
Yes, Thanks!!!
Polymorphic relationship.... Now I am confident that you are getting my point.

Thanks!!!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
